backends/fs/morphos/morphos-fs.cpp

```diff
@@ -0,0 +1,375 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ *
+ */
+
+#if defined(__MORPHOS__)
+
+#include "backends/fs/morphos/morphos-fs.h"
+#include "backends/fs/stdiostream.h"
+#include "common/debug.h"
+#include "common/util.h"
+
+/**
+ * Returns the last component of a given path.
+ *
+ * @param str Common::String containing the path.
+ * @return Pointer to the first char of the last component inside str.
+ */
+const char *lastPathComponent(const Common::String &str) {
+	int offset = str.size();
+
+	if (offset <= 0) {
+		debug(6, "Bad offset");
+		return 0;
+	}
+
+	const char *p = str.c_str();
+
+	while (offset > 0 && (p[offset-1] == '/' || p[offset-1] == ':'))
+		offset--;
+
+	while (offset > 0 && (p[offset-1] != '/' && p[offset-1] != ':'))
+		offset--;
+
+	return p + offset;
+}
+
+MorphOSFilesystemNode::MorphOSFilesystemNode() {
+
+	_sDisplayName = "Available HDDs/Partitions";
+	_bIsValid = true;
+	_bIsDirectory = true;
+	_sPath = "";
+	_pFileLock = 0;
+	_nProt = 0; // Protection is ignored for the root volume
+
+}
+
+MorphOSFilesystemNode::MorphOSFilesystemNode(const Common::String &p) {
+
+	int offset = p.size();
+
+	if (offset <= 0) {
+		debug(6, "Bad offset");
+		return;
+	}
+
+	_sPath = p;
+	_sDisplayName = ::lastPathComponent(_sPath);
+	_pFileLock = 0;
+	_bIsDirectory = false;
+	_bIsValid = false;
+
+	struct FileInfoBlock *fib = (struct FileInfoBlock*) AllocDosObject(DOS_FIB, NULL);
+	
+	if (fib == NULL) {
+	 debug(6,"Failed...");
+	 return;	
+	}
+
+	BPTR pLock = Lock((STRPTR)_sPath.c_str(), SHARED_LOCK);
+	if (pLock) {
+	 if (Examine(pLock, fib) != DOSFALSE) {
+	  if (fib->fib_EntryType > 0)
+	  {
+	   	_bIsDirectory = true;
+	   	_pFileLock = DupLock(pLock);
+		_bIsValid = (_pFileLock != 0);
+	    const char c = _sPath.lastChar();
+			if (c != '/' && c != ':')
+				_sPath += '/';
+	  } else {
+	  	_bIsDirectory = false;
+		_bIsValid = false;
+	  }
+	 }
+	 UnLock(pLock);
+	}
+	FreeDosObject(DOS_FIB, fib);
+}
+
+MorphOSFilesystemNode::MorphOSFilesystemNode(BPTR pLock, const char *pDisplayName) {
+
+	int bufSize = MAXPATHLEN;
+	_pFileLock = 0;
+
+	while (true) {
+		char *n = new char[bufSize];
+		if (NameFromLock(pLock, (STRPTR)n, bufSize) != DOSFALSE) {
+			_sPath = n;
+			_sDisplayName = pDisplayName ? pDisplayName : FilePart((STRPTR)n);
+			delete[] n;
+			break;
+		}
+
+		if (IoErr() != ERROR_LINE_TOO_LONG) {
+			_bIsValid = false;
+			debug(6, "IoErr() != ERROR_LINE_TOO_LONG");
+			delete[] n;
+			return;
+		}
+
+		bufSize *= 2;
+		delete[] n;
+	}
+
+	_bIsValid = false;
+	_bIsDirectory = false;
+
+	FileInfoBlock *fib = (FileInfoBlock*) AllocDosObject(DOS_FIB, NULL);
+	
+	if (fib == NULL) {
+	 debug(6,"Failed...");
+	 return;	
+	}
+
+	if (Examine(pLock, fib) != DOSFALSE) 
+	{
```

Comment:
Code formatting is wrong here. The braces must stick to the previous line.

Reply:
hi, ok i reformat that.